harness/determined/searcher/search_method.py

```diff
@@ -0,0 +1,245 @@
+        super().__init__()
+        self.request_id = request_id
+
+    def _to_searcher_operation(self) -> bindings.v1SearcherOperation:
+        return bindings.v1SearcherOperation(
+            closeTrial=bindings.v1CloseTrialOperation(requestId=str(self.request_id))
+        )
+
+
+class Progress(Operation):
+    def __init__(self, progress: float):
+        super().__init__()
+        self.progress = progress
+
+    def _to_searcher_operation(self) -> bindings.v1SearcherOperation:
+        return bindings.v1SearcherOperation(
+            searcherProgress=bindings.v1SearcherProgressOperation(progress=self.progress)
+        )
+
+
+class Shutdown(Operation):
+    def __init__(self) -> None:
+        super().__init__()
+
+    def _to_searcher_operation(self) -> bindings.v1SearcherOperation:
+        return bindings.v1SearcherOperation(shutdown=bindings.v1ShutdownOperation())
+
+
+class Create(Operation):
+    def __init__(
+        self,
+        request_id: uuid.UUID,
+        hparams: Dict[str, Any],
+        checkpoint: Optional[Checkpoint],
+    ) -> None:
+        super().__init__()
+        self.request_id = request_id
+        self.hparams = json.dumps(hparams)
+        self.checkpoint = checkpoint
+
+    def _to_searcher_operation(self) -> bindings.v1SearcherOperation:
+        return bindings.v1SearcherOperation(
+            createTrial=bindings.v1CreateTrialOperation(
+                hyperparams=self.hparams, requestId=str(self.request_id)
+            )
+        )
+
+
+class SearchMethod:
+    def __init__(self) -> None:
+        self._searcher_state = SearcherState()
+
+    @property
+    def searcher_state(self) -> SearcherState:
+        return self._searcher_state
+
+    @abstractmethod
+    def initial_operations(self) -> List[Operation]:
+        """
+        initial_operations returns a set of initial operations that the searcher
+        would like to take.
+        """
+        pass
+
+    @abstractmethod
+    def on_trial_created(self, request_id: uuid.UUID) -> List[Operation]:
+        """
+        on_trial_created informs the searcher that a trial has been created
+        as a result of Create operation.
+        """
+        pass
+
+    @abstractmethod
+    def on_validation_completed(self, request_id: uuid.UUID, metric: float) -> List[Operation]:
+        """
+        on_validation_completed informs the searcher that the validation workload
+        initiated by the same searcher has completed. It returns any new operations
+        as a result of this workload completing.
+        """
+        pass
+
+    @abstractmethod
+    def on_trial_closed(self, request_id: uuid.UUID) -> List[Operation]:
+        """
+        trialClosed informs the searcher that the trial has been closed as a result of a Close
+        operation.
+        """
+        pass
+
+    @abstractmethod
+    def progress(self) -> float:
+        """
+        progress returns experiment progress as a float between 0 and 1.
+        """
+        pass
```

Comment:
Do we synthesize progress events in the master when operations are completed?

If we don't... report_progress() is actually an optional call so I want to make sure that report_progress() isn't the only way this gets called.

Separately, other events can still emit Progress operations, right?  Why not just let Progress be a normal operation that can be emitted anytime, and the progress event just returns normal operations as well?

Reply:
> Do we synthesize progress events in the master when operations are completed?

Not sure what you mean by synthesizing. `report_progress` _is_ the only way this gets called. I don't see a problem with that.

Progress operations are not user-facing (maybe I should actually rename `Progress -> _Progress`), they have been introduced by us here solely as a vehicle to convey experiment progress back to master after another trial reports progress to master and master tells the SearchRunner about it. The user should not be aware that Progress is an operation and they should not mix Progress into a list of operations returned by other event handlers.

We can discuss this.

Reply:
This will be addressed in https://github.com/determined-ai/determined/pull/5158 which I will merge once the build is successful.